File: programmers/42626.cpp
#include <string>
#include <vector>
using namespace std;

void heap_push(vector<int> &heap, int value)
{
    heap.push_back(value);
    int current = heap.size() - 1;
    while (current > 0)
    {
        int parent_pos = (current - 1) >> 1;
        int parent = heap[parent_pos];
        if (value < parent)
        {
            heap[current] = parent;
            current = parent_pos;
            continue;
        }
        break;
    }
    heap[current] = value;
}

void fix_heap_pop(vector<int> &heap)
{
    int oldest = heap[0];
    int current = 0;
    int length = heap.size();
    int left_pos = 2 * current + 1;
    int right_pos = 2 * current + 2;
    if (right_pos == length)
    {
        if (heap[0] > heap[1])
        {
            swap(heap[0], heap[1]);
            return;
        }
    }

    while (right_pos < length)
    {
        if (oldest < heap[left_pos] and oldest < heap[right_pos])
            break;
        else if (heap[left_pos] < heap[right_pos])
        {
            heap[current] = heap[left_pos];
            current = left_pos;
        }

        else
        {
            heap[current] = heap[right_pos];
            current = right_pos;
        }

        left_pos = 2 * current + 1;
        right_pos = 2 * current + 2;
    }

    heap[current] = oldest;
}

int heap_pop(vector<int> &heap)
{
    int oldest = heap.back();
    heap.pop_back();
    if (heap.empty())
        return oldest;
    int value = heap[0];
    heap[0] = oldest;
    fix_heap_pop(heap);
    return value;
}

int solution(vector<int> scoville, int K)
{
    vector<int> heap;
    for (int i = 0; i < scoville.size(); i++)
    {
        heap_push(heap, scoville[i]);
    }
    scoville = heap;

    int answer = 0;
    while (scoville[0] < K)
    {
        if (scoville.size() < 2)
            return -1;
        int min_one = heap_pop(scoville);
        int min_two = heap_pop(scoville);
        int mixed = min_one + min_two * 2;
        heap_push(scoville, mixed);
        answer += 1;
    }
    return answer;
}

int main()
{
    vector<int> scoville;
    scoville.push_back(1);
    scoville.push_back(2);
    scoville.push_back(3);
    scoville.push_back(9);
    scoville.push_back(10);
    scoville.push_back(12);
    int K = 34;
    solution(scoville, K);
}

/*
효율성  테스트
테스트 1 〉	통과 (21.95ms, 9.35MB)
테스트 2 〉	통과 (46.92ms, 14.9MB)
테스트 3 〉	통과 (214.67ms, 39.7MB)
테스트 4 〉	통과 (17.02ms, 8MB)
테스트 5 〉	통과 (229.15ms, 41.3MB
*/

File: programmers/42626.py
MAX=1000000001

def min_heapify(arr):
    heap = list()
    for v in arr:
        heap = heap_push(heap, v)
    return heap

def heap_push(heap, value):
    heap.append(value)
    current = len(heap)-1
    while(current > 0 and heap[current] < heap[(current-1)//2]):
        heap[current], heap[(current-1)//2] = heap[(current-1)//2], heap[current]
        current = (current-1)//2
    return heap

def heap_pop(heap):
    value = heap[0]
    length = len(heap) - 1
    heap[0] = heap[length]
    heap[length] = MAX    
    current = 0
    while (2*current+1 < length):
        if heap[current] < heap[2*current+1] and heap[current] < heap[2*current+2]:
            break;
        elif heap[2*current+2] > heap[2*current+1]:
            heap[current],heap[2*current+1] = heap[2*current+1],heap[current]
            current = 2* current + 1
        else:
            heap[current],heap[2*current+2] = heap[2*current+2],heap[current]
            current = 2* current + 2
    return (value, heap[:length])

def solution(scoville, K):
    heap = min_heapify(scoville)
    answer = 0
    while len(heap) > 1:
        min_one, heap = heap_pop(heap)
        min_two, heap = heap_pop(heap)
        mixed = min_one + min_two * 2
        heap = heap_push(heap, mixed)
        answer += 1
        
        if K <= heap[0]:
            break    
    
    if K > heap[0]:
        return -1
    return answer

